File: PART_5/Recursive.java
package PART_5;

public class Recursive {
    public static class Node{
        private Node next;
        private int value;
        public Node(int value){
            this.value = value;
            this.next = null;
        }
    }
    public static class BNode{
        private BNode left;
        private BNode right;
        private int value;
        public BNode(int value){
            this.value = value;
            this.left = null;
            this.right = null;
        }
        public BNode getRight(){
            return this.right;
        }
        public BNode getLeft(){
            return this.left;
        }
        public int getValue(){
            return this.value;
        }
        public void setLeft(int value){
            this.left = new BNode(value);
        }
        public void setRight(int value){
            this.right = new BNode(value);
        }
    }
    public static class BTree{
        private BNode root;
        public BTree(){
            this.root = null;
        }
        private void recursive_insert(BNode node, int value){
            if(node.getValue() >= value){
                if(node.getLeft() == null){
                    node.setLeft(value);
                }else{
                    recursive_insert(node.getLeft(), value);
                }
            }else{
                if(node.getRight() == null){
                    node.setRight(value);
                }else{
                    recursive_insert(node.getRight(), value);
                }
            }
        }
        private void recursive_in_order(BNode node){
            if(node != null){
                recursive_in_order(node.getLeft());
                System.out.print(node.getValue() + "\t");
                recursive_in_order(node.getRight());
            }
        }
        public void in_order(){
            if(this.root == null){
                System.out.println("Empty Btree");
                return;
            }
            recursive_in_order(this.root);
        }
        public void insert(int value){
            if(this.root == null){
                this.root = new BNode(value);
                return;
            }
            recursive_insert(root, value);
        }
    }
    public static class LinkedList{
        private Node head;
        private int lenght;
        public LinkedList(){
            this.head = null;
            this.lenght = 0;
        }
        public void push(int value){
            if(this.head == null){
                this.head = new Node(value);
                this.lenght++;
                return;
            }
            Node current = this.head;
            while(current.next != null){
                current = current.next;
            }
            current.next = new Node(value);
        }
        public void traverse(){
            if(this.head.equals(null)){
                System.out.println("Empty Linked List");
                return;
            }
            Node current = this.head;
            while(current != null){
                System.out.print(current.value + "--->");
                current = current.next;
            }
            System.out.println("null");
        }

        public static void main(String[] args) {
            // Linked List
            LinkedList list = new LinkedList();
            list.push(10);
            list.push(20);
            list.push(30);
            list.traverse();
            // Binary Tree
            BTree tree = new BTree();
            tree.insert(10);
            tree.insert(-4);
            tree.insert(11);
            tree.insert(24);
            tree.insert(-6);
            tree.in_order();
        }
    }
}

File: PART_5/Recursive.cs

using System;


public class Workshop
{
    public class ListNode
    {
        public int Value;
        public ListNode Next;
        public ListNode(int value)
        {
            Value = value;
            Next = null;
        }
    }

    public class LinkedList
    {
        public ListNode Head;
        public int Length;
        public LinkedList()
        {
            Head = null;
            Length = 0;
        }
        public void Add(int value)
        {
            if (Head == null)
            {
                Head = new ListNode(value);
                Length++;
                return;
            }
            ListNode current = Head;
            while (current.Next != null)
            {
                current = current.Next;
            }
            current.Next = new ListNode(value);
            Length++;
        }
        public void Traverse()
        {
            if (Head == null)
            {
                Console.WriteLine("Linked list is empty");
                return;
            }
            ListNode current = Head;
            while (current != null)
            {
                Console.Write(current.Value + "--->");
                current = current.Next;
            }
            Console.WriteLine("null");
        }
    }

    public class TreeNode
    {
        public int Value;
        public TreeNode Left;
        public TreeNode Right;
        public TreeNode(int value)
        {
            Value = value;
            Left = null;
            Right = null;
        }
    }

    public class BinaryTree
    {
        public TreeNode Root;
        public BinaryTree()
        {
            Root = null;
        }
        public void Insert(int value)
        {
            if (Root == null)
            {
                Root = new TreeNode(value);
                return;
            }
            InsertRecursive(Root, value);
        }
        private void InsertRecursive(TreeNode node, int value)
        {
            if (value < node.Value)
            {
                if (node.Left == null)
                {
                    node.Left = new TreeNode(value);
                }
                else
                {
                    InsertRecursive(node.Left, value);
                }
            }
            else
            {
                if (node.Right == null)
                {
                    node.Right = new TreeNode(value);
                }
                else
                {
                    InsertRecursive(node.Right, value);
                }
            }
        }
        public void InOrderTraversal()
        {
            InOrderRecursive(Root);
        }
        private void InOrderRecursive(TreeNode node)
        {
            if (node != null)
            {
                InOrderRecursive(node.Left);
                Console.Write(node.Value + " ");
                InOrderRecursive(node.Right);
            }
        }
    }

    public static void Main(string[] args)
    {
        LinkedList list = new LinkedList();
        list.Add(10);
        list.Add(20);
        list.Add(30);
        Console.WriteLine("Linked list:");
        list.Traverse();

        BinaryTree tree = new BinaryTree();
        tree.Insert(10);
        tree.Insert(-4);
        tree.Insert(11);
        tree.Insert(24);
        tree.Insert(-6);
        Console.WriteLine("\nBinary tree in order:");
        tree.InOrderTraversal();
        Console.WriteLine();
    }
}